cmd/soroban-rpc/internal/methods/get_events.go

```diff
@@ -0,0 +1,475 @@
+}
+
+func (g *GetEventsRequest) Valid() error {
+	// Validate start & end ledger
+	// Validate the ledger range min/max
+	if g.EndLedger < g.StartLedger {
+		return errors.New("endLedger must be after or the same as startLedger")
+	}
+	if g.EndLedger-g.StartLedger > MAX_LEDGER_RANGE {
+		return fmt.Errorf("endLedger must be less than %d ledgers after startLedger", MAX_LEDGER_RANGE)
+	}
+
+	// Validate filters
+	if len(g.Filters) > 5 {
+		return errors.New("maximum 5 filters per request")
+	}
+	for i, filter := range g.Filters {
+		if err := filter.Valid(); err != nil {
+			return errors.Wrapf(err, "filter %d invalid", i+1)
+		}
+	}
+
+	return nil
+}
+
+func (g *GetEventsRequest) Matches(event xdr.ContractEvent) bool {
+	if len(g.Filters) == 0 {
+		return true
+	}
+	for _, filter := range g.Filters {
+		if filter.Matches(event) {
+			return true
+		}
+	}
+	return false
+}
+
+type EventFilter struct {
+	EventType   string        `json:"type,omitempty"`
+	ContractIDs []string      `json:"contractIds,omitempty"`
+	Topics      []TopicFilter `json:"topics,omitempty"`
+}
+
+func (e *EventFilter) Valid() error {
+	switch e.EventType {
+	case "", "system", "contract":
+		// ok
+	default:
+		return errors.New("if set, type must be either 'system' or 'contract'")
+	}
+	if len(e.ContractIDs) > 5 {
+		return errors.New("maximum 5 contract IDs per filter")
+	}
+	if len(e.Topics) > 5 {
+		return errors.New("maximum 5 topics per filter")
+	}
+	for i, id := range e.ContractIDs {
+		out, err := hex.DecodeString(id)
+		if err != nil || len(out) != 32 {
+			return fmt.Errorf("contract ID %d invalid", i+1)
+		}
+	}
+	for i, topic := range e.Topics {
+		if err := topic.Valid(); err != nil {
+			return errors.Wrapf(err, "topic %d invalid", i+1)
+		}
+	}
+	return nil
+}
+
+// TODO: Implement this more efficiently (ideally do it in the real data backend)
+func (e *EventFilter) Matches(event xdr.ContractEvent) bool {
+	return e.matchesEventType(event) && e.matchesContractIDs(event) && e.matchesTopics(event)
+}
+
+func (e *EventFilter) matchesEventType(event xdr.ContractEvent) bool {
+	if e.EventType == "contract" && event.Type != xdr.ContractEventTypeContract {
+		return false
+	}
+	if e.EventType == "system" && event.Type != xdr.ContractEventTypeSystem {
+		return false
+	}
+	return true
+}
+
+func (e *EventFilter) matchesContractIDs(event xdr.ContractEvent) bool {
+	if len(e.ContractIDs) == 0 {
+		return true
+	}
+	if event.ContractId == nil {
+		return false
+	}
+	needle := hex.EncodeToString((*event.ContractId)[:])
+	for _, id := range e.ContractIDs {
+		if id == needle {
+			return true
+		}
+	}
+	return false
+}
+
+func (e *EventFilter) matchesTopics(event xdr.ContractEvent) bool {
+	if len(e.Topics) == 0 {
+		return true
+	}
+	v0 := event.Body.MustV0()
+	for _, topicFilter := range e.Topics {
+		if topicFilter.Matches(v0.Topics) {
+			return true
+		}
+	}
+	return false
+}
+
+type TopicFilter []SegmentFilter
+
+func (t *TopicFilter) Valid() error {
```

Comment:
should this also inspect the filter for the wildcard validity, so could remove the repetitive checks on filter validity and ensuing `panic` in `TopicFilter.Matches` ?